multicluster/cmd/service-mirror/main.go

```diff
@@ -84,79 +103,159 @@ func Main(args []string) {
 	}
 
 	linkClient := k8sAPI.DynamicClient.Resource(multicluster.LinkGVR).Namespace(*namespace)
-
 	metrics := servicemirror.NewProbeMetricVecs()
-
 	controllerK8sAPI.Sync(nil)
 
 	ready = true
-
-main:
-	for {
-		// Start link watch
-		linkWatch, err := linkClient.Watch(ctx, metav1.ListOptions{})
-		if err != nil {
-			log.Fatalf("Failed to watch Link %s: %s", linkName, err)
-		}
-		results := linkWatch.ResultChan()
-
-		// Each time the link resource is updated, reload the config and restart the
-		// cluster watcher.
+	run := func(ctx context.Context) {
 		for {
-			select {
-			case <-stop:
-				break main
-			case event, ok := <-results:
-				if !ok {
-					log.Info("Link watch terminated; restarting watch")
-					continue main
-				}
-				switch obj := event.Object.(type) {
-				case *dynamic.Unstructured:
-					if obj.GetName() == linkName {
-						switch event.Type {
-						case watch.Added, watch.Modified:
-							link, err := multicluster.NewLink(*obj)
-							if err != nil {
-								log.Errorf("Failed to parse link %s: %s", linkName, err)
-								continue
-							}
-							log.Infof("Got updated link %s: %+v", linkName, link)
-							creds, err := loadCredentials(ctx, link, *namespace, k8sAPI)
-							if err != nil {
-								log.Errorf("Failed to load remote cluster credentials: %s", err)
-							}
-							err = restartClusterWatcher(ctx, link, *namespace, creds, controllerK8sAPI, *requeueLimit, *repairPeriod, metrics, *enableHeadlessSvc)
-							if err != nil {
-								// failed to restart cluster watcher; give a bit of slack
-								// and restart the link watch to give it another try
-								log.Error(err)
-								time.Sleep(linkWatchRestartAfter)
-								linkWatch.Stop()
-							}
-						case watch.Deleted:
-							log.Infof("Link %s deleted", linkName)
-							if clusterWatcher != nil {
-								clusterWatcher.Stop(false)
-								clusterWatcher = nil
-							}
-							if probeWorker != nil {
-								probeWorker.Stop()
-								probeWorker = nil
+			// Start link watch
+			linkWatch, err := linkClient.Watch(ctx, metav1.ListOptions{})
+			if err != nil {
+				log.Fatalf("Failed to watch Link %s: %s", linkName, err)
+			}
+			results := linkWatch.ResultChan()
+
+			// Each time the link resource is updated, reload the config and restart the
+			// cluster watcher.
+			for loop := true; loop; {
+				select {
+				// ctx.Done() is a one-shot channel that will be closed once
+				// the context has been cancelled. Receiving from a closed
+				// channel yields the value immediately.
+				case <-ctx.Done():
+					// The channel will be closed by the leader elector when a
+					// lease is lost, or by a background task handling SIGTERM.
+					// Before terminating the loop, stop the workers and set
+					// them to nil to release memory.
+					cleanupWorkers()
+					return
+				case event, ok := <-results:
+					if !ok {
+						log.Info("Link watch terminated; restarting watch")
+						loop = false
```

Comment:
this used to be a named continue.  any specific reason to change it?

Reply:
Just preference on my side. I also did not like that loop labels were indented too much to the left. Since the code is easier to understand using loop labels, I've added them back in.